### DISCUSSION TITLE
Override / patch user c module function in python

### DISCUSSION THREAD

**Discussion by schnili** [micropython/micropython#15885] [General]
Hey,
I have written a c module like described in the [documentation](https://docs.micropython.org/en/latest/develop/library.html).
Is it possible to allow overwriting a function in it (e.g. monkey patching) afterwards?

e.g.
```python
import mycmodule
def test():
    print("hi")
mycmodule.func = test
mycmodule.func()
```
Running this with the c module gets me `AttributeError: 'module' object has no attribute func`.

I found the `MICROPY_CAN_OVERRIDE_BUILTINS` option. But I dont quite get how to use it / if it even applies to my usecase.

Thanks!

**Comment by stinos** [comment]
IIRC you need to implement setting attributes for your module using `MP_REGISTER_MODULE_DELEGATION`; can probably grep the code base for sample code.

**Reply by schnili** [reply]
Thanks for the hint!
I'll have to look into that, its used in the webassembly/js port and in modsys, need to figure out the API :)

**Reply by schnili** [reply]
I got it working :+1: :
```c
[...]
STATIC const uint16_t mycmodule_mutable_keys[] = {
    MP_QSTR_func,
    MP_QSTRnull,
};

MP_REGISTER_ROOT_POINTER(mp_obj_t mycmodule_mutable[2]);

void mp_module_mycmodule_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest)
{
    MP_STATIC_ASSERT(MP_ARRAY_SIZE(MP_STATE_VM(mycmodule_mutable)) ==
                     MP_ARRAY_SIZE(mycmodule_mutable_keys));
    mp_module_generic_attr(attr, dest, mycmodule_mutable_keys, MP_STATE_VM(mycmodule_mutable));
}

MP_REGISTER_MODULE(MP_QSTR_mycmodule, my_cmodule);
MP_REGISTER_MODULE_DELEGATION(my_cmodule, mp_module_mycmodule_attr);
```

**Comment by corranwebster** [comment]
An alternative in pure micropython: wrap your c module in a .py module which simply imports from the c module and then you can monkey-patch away.  ie.

mymodule.py
``` python
from ._mymodule import  func  # import from _mymodule.c
```
Then your example code should work fine.

**Reply by schnili** [reply]
thanks!
I thought about that, but as the c module is quite big I don't want to introduce a shim just for 1 function that i want to define in python

**Comment by shariltumin** [comment]
If I understand you correctly, what you want to achieve is something like this:

```
$ python
Python 3.10.2 (main, Feb  4 2022, 18:51:31) [GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import time
>>> ts=time.sleep
>>> ts(1)
>>> def fun(t):
...     print(f'Sleeping for {t} sec.')
...     ts(t)
... 
>>> time.sleep=fun
>>> time.sleep(2)
Sleeping for 2 sec.
>>> 
```

However, this will cause an "object has no attribute" error in MicroPython.

One solution is to copy all module attributes to a dict, preferably one that supports dot notation.

See "dotdict.py" for a simple implementation:

```py
_B='No such attribute: '
class Dot(dict):
  def __getattr__(B,N):
     if N in B:return B[N]
     else:raise AttributeError(_B+N)
  def __setattr__(B,N,V):
     B[N]=V
  def __delattr__(B,N):
     if N in B:del B[N]
     else:raise AttributeError(_B+N)
```

With this in place, we can then do the same in MicroPython:

```
MicroPython v1.24.0-preview.315.g8feb714b4.dirty on 2024-09-20; ESP32 CLASSIC with ESP32
>>> import time as T
>>> from dotdict import Dot
>>> time=Dot()
>>> time.update(T.__dict__)
>>> ts = time.sleep
>>> def fun(t):
...     print(f'Sleeping for {t} sec.')
...     ts(t)
... 
>>> time.sleep=fun
>>> time.sleep(3)
Sleeping for 3 sec.
>>> 
```

The "dotdict.py" will probably work on all ports, except a port that does not support ```MICROPY_PY_DELATTR_SETATTR``` like the ESP8266-GENERIC.
